syntax = "proto3";

// Client-generated feedback used for system monitoring and planning.
message UserFeedback {
  // Observed latency measurements for individual keys.
  message KeyLatency {
    // The key for which latency is being reported.
    string key = 1;

    // The observed latency for this key.
    double latency = 2;
  }
  
  // A unique ID representing each individual client.
  string uid = 1;

  // Perceived average latency across all requests made by this client.
  double avg_latency = 2;

  // Perceived min latency
  double min_latency = 3;

  // Perceived max latency 
  double max_latency = 4;

  // Perceived median latency
  double median_latency = 5;

  // Perceived tail latency made by this client.
  double tail_latency = 6;

  // Notifies the monitoring system that the running benchmark has finished.
  bool finish = 7;

  // The perceived throughput across all keys.
  double throughput = 8;

  // Set during the benchmark warm-up phase to tell the monitoring system that
  // it should ignore policy decisions.
  bool warmup = 9;

  // Perceived latencies for individual keys.
  repeated KeyLatency key_latency = 10;
}